ddl/ddl_worker.go

```diff
@@ -464,3 +476,64 @@ func updateSchemaVersion(t *meta.Meta, job *model.Job) (int64, error) {
 	err = t.SetSchemaDiff(diff)
 	return schemaVersion, errors.Trace(err)
 }
+
+func (d *ddl) cleanAddIndexQueueJobs(txn kv.Transaction) error {
+	startTime := time.Now()
+	m := meta.NewMeta(txn)
+	m.SetJobListKey(meta.AddIndexJobListKey)
+	for {
+		job, err := d.getFirstDDLJob(m)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if job == nil {
+			log.Infof("[ddl] cleaning jobs in the adding index queue takes time %v.", time.Since(startTime))
+			return nil
+		}
+		log.Infof("[ddl] cleaning job %v in the adding index queue.", job)
+
+		// The types of these jobs must be ActionAddIndex.
+		if job.SchemaState == model.StatePublic || job.SchemaState == model.StateNone {
+			job.State = model.JobStateCancelled
+			err = d.finishDDLJob(m, job)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
+
+		schemaID := job.SchemaID
+		tblInfo, err := getTableInfo(m, job, schemaID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		var indexName model.CIStr
+		var unique bool
+		err = job.DecodeArgs(&unique, &indexName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		indexInfo := findIndexByName(indexName.L, tblInfo.Indices)
+		job.State = model.JobStateRollingback
+		job.Args = []interface{}{indexInfo.Name}
+		// If add index job rollbacks, its work is the same as drop index job do.
+		// When it's not in None state, the next state can be delete only state.
```

Comment:
Maybe this way is more friendly for reader? // When it's not in "None" state, the next state can be "delete only" state.